File: solutions/java/code-carl/ch06_tree/T8_LC0104_MaxDepth.java
/**
 * Practice template for:
 * - LeetCode 104. Maximum Depth of Binary Tree
 *   https://leetcode.cn/problems/maximum-depth-of-binary-tree/
 */
public class T8_LC0104_MaxDepth {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = maxDepth(root.left);
        int right = maxDepth(root.right);
        return Math.max(left, right) + 1;
    }
}

File: solutions/code-carl/ch06_tree/t8_lc0104_max_depth.py
"""LeetCode 104. Maximum Depth of Binary Tree
https://leetcode.cn/problems/maximum-depth-of-binary-tree/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1


if __name__ == "__main__":
    solution = Solution()
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert solution.maxDepth(root) == 3
    assert solution.maxDepth(None) == 0
    print("All tests passed.")
